File: iphone/Woladen/Models/FilterState.swift
import Foundation

struct FilterState: Codable, Equatable {
    var selectedOperatorNames: Set<String> = []
    var minPowerKW: Double = 50
    var minAmenityCount: Double = 0
    var selectedAmenities: Set<String> = []
    var amenityNameQuery: String = ""
    var availableOnly: Bool = true
    var currentlyOpenOnly: Bool = false
    var routeMaxDistanceFromLocationKM: Double?

    enum CodingKeys: String, CodingKey {
        case operatorName
        case selectedOperatorNames
        case minPowerKW
        case minAmenityCount
        case selectedAmenities
        case amenityNameQuery
        case availableOnly
        case currentlyOpenOnly
        case routeMaxDistanceFromLocationKM
    }

    init(
        operatorName: String = "",
        selectedOperatorNames: Set<String> = [],
        minPowerKW: Double = 50,
        minAmenityCount: Double = 0,
        selectedAmenities: Set<String> = [],
        amenityNameQuery: String = "",
        availableOnly: Bool = true,
        currentlyOpenOnly: Bool = false,
        routeMaxDistanceFromLocationKM: Double? = nil
    ) {
        let selectedNames = Self.normalizedOperatorNames(selectedOperatorNames)
        self.selectedOperatorNames = selectedNames.isEmpty
            ? Self.normalizedOperatorNames([operatorName])
            : selectedNames
        self.minPowerKW = minPowerKW
        self.minAmenityCount = minAmenityCount
        self.selectedAmenities = selectedAmenities
        self.amenityNameQuery = amenityNameQuery
        self.availableOnly = availableOnly
        self.currentlyOpenOnly = currentlyOpenOnly
        self.routeMaxDistanceFromLocationKM = Self.normalizedRouteDistance(routeMaxDistanceFromLocationKM)
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let decodedOperatorNames = (try? container.decode(Set<String>.self, forKey: .selectedOperatorNames)) ?? []
        let legacyOperatorName = (try? container.decode(String.self, forKey: .operatorName)) ?? ""
        let selectedNames = Self.normalizedOperatorNames(decodedOperatorNames)
        selectedOperatorNames = selectedNames.isEmpty
            ? Self.normalizedOperatorNames([legacyOperatorName])
            : selectedNames
        minPowerKW = (try? container.decode(Double.self, forKey: .minPowerKW)) ?? 50
        minAmenityCount = (try? container.decode(Double.self, forKey: .minAmenityCount)) ?? 0
        selectedAmenities = (try? container.decode(Set<String>.self, forKey: .selectedAmenities)) ?? []
        amenityNameQuery = (try? container.decode(String.self, forKey: .amenityNameQuery)) ?? ""
        availableOnly = (try? container.decode(Bool.self, forKey: .availableOnly)) ?? true
        currentlyOpenOnly = (try? container.decode(Bool.self, forKey: .currentlyOpenOnly)) ?? false
        routeMaxDistanceFromLocationKM = Self.normalizedRouteDistance(
            try? container.decode(Double.self, forKey: .routeMaxDistanceFromLocationKM)
        )
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(selectedOperatorNames.sorted(), forKey: .selectedOperatorNames)
        try container.encode(operatorName, forKey: .operatorName)
        try container.encode(minPowerKW, forKey: .minPowerKW)
        try container.encode(minAmenityCount, forKey: .minAmenityCount)
        try container.encode(selectedAmenities.sorted(), forKey: .selectedAmenities)
        try container.encode(amenityNameQuery, forKey: .amenityNameQuery)
        try container.encode(availableOnly, forKey: .availableOnly)
        try container.encode(currentlyOpenOnly, forKey: .currentlyOpenOnly)
        try container.encodeIfPresent(routeMaxDistanceFromLocationKM, forKey: .routeMaxDistanceFromLocationKM)
    }

    var operatorName: String {
        get {
            selectedOperatorNames.count == 1 ? selectedOperatorNames.sorted().first ?? "" : ""
        }
        set {
            selectedOperatorNames = Self.normalizedOperatorNames([newValue])
        }
    }

    var activeCount: Int {
        var count = 0
        if !selectedOperatorNames.isEmpty { count += 1 }
        if minPowerKW > 0 { count += 1 }
        if minAmenityCount > 0 { count += 1 }
        count += selectedAmenities.count
        if !amenityNameQuery.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty { count += 1 }
        if availableOnly { count += 1 }
        if currentlyOpenOnly { count += 1 }
        if routeMaxDistanceFromLocationKM != nil { count += 1 }
        return count
    }

    var clearableState: FilterState {
        FilterState(
            selectedOperatorNames: [],
            minPowerKW: 50,
            minAmenityCount: 0,
            selectedAmenities: [],
            amenityNameQuery: "",
            availableOnly: false,
            currentlyOpenOnly: false,
            routeMaxDistanceFromLocationKM: nil
        )
    }

    private static func normalizedOperatorNames(_ values: Set<String>) -> Set<String> {
        Set(
            values
                .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
                .filter { !$0.isEmpty }
        )
    }

    private static func normalizedRouteDistance(_ value: Double?) -> Double? {
        guard let value, value.isFinite, value > 0 else { return nil }
        return min(value, 400)
    }
}

enum FilterStateStore {
    private static let key = "woladen.filterState.v1"

    static func load() -> FilterState {
        guard let data = UserDefaults.standard.data(forKey: key) else {
            return FilterState()
        }
        return (try? JSONDecoder().decode(FilterState.self, from: data)) ?? FilterState()
    }

    static func save(_ state: FilterState) {
        guard let data = try? JSONEncoder().encode(state) else { return }
        UserDefaults.standard.set(data, forKey: key)
    }
}

File: android/app/src/main/java/de/woladen/android/model/FilterState.kt
package de.woladen.android.model

data class FilterState(
    val selectedOperatorNames: Set<String> = emptySet(),
    val minPowerKw: Double = 50.0,
    val minAmenityCount: Double = 0.0,
    val selectedAmenities: Set<String> = emptySet(),
    val amenityNameQuery: String = "",
    val availableOnly: Boolean = true,
    val currentlyOpenOnly: Boolean = false,
    val routeMaxDistanceFromLocationKm: Double? = null
) {
    fun canonicalized(using: List<OperatorEntry>): FilterState {
        if (normalizedOperatorNames.isEmpty() || using.isEmpty()) return this
        val migrated = normalizedOperatorNames.mapNotNull { selected ->
            val exactID = using.firstOrNull { it.id.equals(selected, ignoreCase = true) }
            val namedEntry = using.firstOrNull { entry ->
                listOf(entry.name).plus(entry.aliases)
                    .any { candidate -> candidate.trim().equals(selected, ignoreCase = true) }
            }
            (exactID ?: namedEntry)?.id
        }.toSet()
        return if (migrated == normalizedOperatorNames) this else copy(selectedOperatorNames = migrated)
    }

    val operatorName: String
        get() = normalizedOperatorNames.singleOrNull().orEmpty()

    val normalizedOperatorNames: Set<String>
        get() = selectedOperatorNames
            .map { it.trim() }
            .filter { it.isNotBlank() }
            .toSet()

    val activeCount: Int
        get() {
            var count = 0
            if (normalizedOperatorNames.isNotEmpty()) count += 1
            if (minPowerKw > 0.0) count += 1
            if (minAmenityCount > 0.0) count += 1
            count += selectedAmenities.size
            if (amenityNameQuery.isNotBlank()) count += 1
            if (availableOnly) count += 1
            if (currentlyOpenOnly) count += 1
            if (routeMaxDistanceFromLocationKm != null) count += 1
            return count
        }

    val clearableState: FilterState
        get() = FilterState(
            selectedOperatorNames = emptySet(),
            minPowerKw = 50.0,
            minAmenityCount = 0.0,
            selectedAmenities = emptySet(),
            amenityNameQuery = "",
            availableOnly = false,
            currentlyOpenOnly = false,
            routeMaxDistanceFromLocationKm = null
        )
}
